autoscale-scaler-kubernetes/src/main/java/com/github/autoscaler/scaler/kubernetes/K8sHealthCheck.java

```diff
@@ -0,0 +1,92 @@
+ */
+package com.github.autoscaler.scaler.kubernetes;
+
+import com.hpe.caf.api.HealthResult;
+import com.hpe.caf.api.HealthStatus;
+import io.kubernetes.client.extended.kubectl.Kubectl;
+import io.kubernetes.client.extended.kubectl.exception.KubectlException;
+import io.kubernetes.client.openapi.ApiException;
+import io.kubernetes.client.openapi.apis.AuthorizationV1Api;
+import io.kubernetes.client.openapi.models.V1ResourceAttributes;
+import io.kubernetes.client.openapi.models.V1SelfSubjectAccessReviewSpec;
+import io.kubernetes.client.openapi.models.V1SelfSubjectAccessReview;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+final class K8sHealthCheck
+{
+    private static final Logger LOG = LoggerFactory.getLogger(K8sHealthCheck.class);
+
+    private K8sHealthCheck()
+    {
+
+    }
+
+    public static HealthResult healthCheck()
+    {
+        final HealthResult connectionHealthResult = checkConnection();
+        if (connectionHealthResult == HealthResult.RESULT_HEALTHY) {
+            return checkPermissions();
+        } else {
+            return connectionHealthResult;
+        }
+    }
+
+    private static HealthResult checkConnection()
+    {
+        try {
+            Kubectl.version().execute();
+            return HealthResult.RESULT_HEALTHY;
+        } catch (KubectlException e) {
+            LOG.warn("Connection failure to kubernetes", e);
+            return new HealthResult(HealthStatus.UNHEALTHY, "Cannot connect to Kubernetes");
+        }
+    }
+
+    private static HealthResult checkPermissions()
+    {
+        final V1ResourceAttributes resourceAttributes = new V1ResourceAttributes();
+        resourceAttributes.setGroup("apps");
+        resourceAttributes.setResource("deployments");
+        resourceAttributes.setVerb("patch");
+        resourceAttributes.setNamespace("private");
+
+        final V1SelfSubjectAccessReviewSpec spec = new V1SelfSubjectAccessReviewSpec();
+        spec.setResourceAttributes(resourceAttributes);
+
+        final V1SelfSubjectAccessReview body = new V1SelfSubjectAccessReview();
+        body.setApiVersion("authorization.k8s.io/v1");
+        body.setKind("SelfSubjectAccessReview");
+        body.setSpec(spec);
+
+        final V1SelfSubjectAccessReview review;
+        try {
+            review = new AuthorizationV1Api().createSelfSubjectAccessReview(body, "All", "fas", "true");
+        } catch (final ApiException e) {
+            throw new RuntimeException(e);
+        }
+
+        if (review.getStatus() != null && review.getStatus().getAllowed()) {
+            return HealthResult.RESULT_HEALTHY;
+        } else {
+            final String errorMessage = String.format("Error: Kubernetes Service Account %s does not have %s permissions to resource %s",
+                    review.getMetadata().getManagedFields().get(0).getManager(), review.getSpec().getResourceAttributes().getVerb(), review.getSpec().getResourceAttributes().getResource());
```

Comment:
Line goes over 139 char limit.

I think we should also include the reason + evaluation error in the log message, and also include the resource  attributes.

Maybe something like:

```
final String serviceAccount = review.getMetadata().getManagedFields().get(0).getManager();
final String evaluationError = review.getStatus().getEvaluationError();
final String reason = review.getStatus().getReason();
final String errorMessage = String.format(
    "Error: Kubernetes Service Account %s does not have permissions to resource %s due to evaluation error %s and reason %s",
    serviceAccount, resourceAttributes, evaluationError, reason);
```

Note there is also a lot of null dereference errors in the above snippet, so you'll probably want to check for nulls on each of these objects

Reply:
It may be simpler just to include the full review object in the log as you were doing originally, this avoids all that null checking etc and makes sure we include anything of relevance in the log

Reply:
@rorytorneymf no problem.
I had committed those changes yesterday evening just to see what it was like. But will just keep in the full review object as it was before. Thanks